File: client/cpp/synnax/framer/writer.cpp
// Copyright 2024 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

/// std
#include <string>


/// api protos
#include "v1/framer.pb.h"

/// internal
#include "synnax/framer/framer.h"

const std::string WRITE_ENDPOINT = "/frame/write";

using namespace synnax;

/// @brief enumeration of possible writer commands.
enum WriterCommand : uint32_t {
    OPEN = 0,
    WRITE = 1,
    COMMIT = 2,
    ERROR = 3,
    SET_AUTHORITY = 4
};


std::pair<Writer, freighter::Error> FrameClient::openWriter(const WriterConfig &config) {
    auto [s, exc] = writer_client->stream(WRITE_ENDPOINT);
    if (exc) return {Writer(), exc};
    auto req = api::v1::FrameWriterRequest();
    req.set_command(OPEN);
    config.toProto(req.mutable_config());
    exc = s->send(req);
    if (exc) return {Writer(), exc};
    auto [_, recExc] = s->receive();
    return {Writer(std::move(s)), recExc};
}

Writer::Writer(std::unique_ptr<WriterStream> s): stream(std::move(s)) {}


void WriterConfig::toProto(api::v1::FrameWriterConfig *f) const {
    subject.to_proto(f->mutable_control_subject());
    f->set_start(start.value);
    for (auto &auth: authorities) f->add_authorities(auth);
    for (auto &ch: channels) f->add_keys(ch);
}

bool Writer::write(Frame fr) {
    assertOpen();
    if (err_accumulated) return false;
    api::v1::FrameWriterRequest req;
    req.set_command(WRITE);
    fr.toProto(req.mutable_frame());
    auto exc = stream->send(req);
    if (exc) err_accumulated = true;
    return !err_accumulated;
}

std::pair<synnax::TimeStamp, bool> Writer::commit() {
    assertOpen();
    if (err_accumulated) return {synnax::TimeStamp(), false};

    auto req = api::v1::FrameWriterRequest();
    req.set_command(COMMIT);
    auto exc = stream->send(req);
    if (exc) {
        err_accumulated = true;
        return {synnax::TimeStamp(0), false};
    }

    while (true) {
        auto [res, recExc] = stream->receive();
        if (recExc) {
            err_accumulated = true;
            return {synnax::TimeStamp(0), false};
        }
        if (res.command() == COMMIT) return {synnax::TimeStamp(res.end()), true};
    }
}

freighter::Error Writer::error() {
    assertOpen();

    auto req = api::v1::FrameWriterRequest();
    req.set_command(ERROR);
    auto exc = stream->send(req);
    if (exc) return exc;

    while (true) {
        auto [res, recExc] = stream->receive();
        if (recExc) return recExc;
        if (res.command() == ERROR) return {res.error()};
    }
}

freighter::Error Writer::close() {
    auto exc = stream->closeSend();
    if (exc) return exc;
    auto [_, recExc] = stream->receive();
    if (recExc.type == freighter::EOF_.type) return freighter::NIL;
    return recExc;
}


void Writer::assertOpen() const {
    if (closed)
        throw std::runtime_error("cannot call method on closed writer");
}

File: cesium/writer.go
// Copyright 2023 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package cesium

import (
	"github.com/cockroachdb/errors"
	"github.com/synnaxlabs/x/confluence"
	"github.com/synnaxlabs/x/signal"
	"github.com/synnaxlabs/x/telem"
	"go.uber.org/zap"
)

type Writer struct {
	requests          confluence.Inlet[WriterRequest]
	responses         confluence.Outlet[WriterResponse]
	wg                signal.WaitGroup
	logger            *zap.Logger
	hasAccumulatedErr bool
}

const unexpectedSteamClosure = "[cesium] - unexpected early closure of response stream"

func wrapStreamWriter(internal StreamWriter) *Writer {
	sCtx, _ := signal.Isolated()
	req := confluence.NewStream[WriterRequest](1)
	res := confluence.NewStream[WriterResponse](1)
	internal.InFrom(req)
	internal.OutTo(res)
	internal.Flow(
		sCtx,
		confluence.CloseInletsOnExit(),
	)
	return &Writer{requests: req, responses: res, wg: sCtx}
}

func (w *Writer) Write(frame Frame) bool {
	if w.hasAccumulatedErr {
		return false
	}
	select {
	case <-w.responses.Outlet():
		w.hasAccumulatedErr = true
		return false
	case w.requests.Inlet() <- WriterRequest{Frame: frame, Command: WriterWrite}:
		return true
	}
}

func (w *Writer) Commit() (telem.TimeStamp, bool) {
	if w.hasAccumulatedErr {
		return 0, false
	}
	select {
	case <-w.responses.Outlet():
		w.hasAccumulatedErr = true
		return 0, false
	case w.requests.Inlet() <- WriterRequest{Command: WriterCommit}:
	}
	for res := range w.responses.Outlet() {
		if res.Command == WriterCommit {
			return res.End, res.Ack
		}
	}
	w.logger.DPanic(unexpectedSteamClosure)
	return 0, false
}

func (w *Writer) Error() error {
	w.requests.Inlet() <- WriterRequest{Command: WriterError}
	for res := range w.responses.Outlet() {
		if res.Command == WriterError {
			w.hasAccumulatedErr = false
			return res.Err
		}
	}
	w.logger.DPanic(unexpectedSteamClosure)
	return errors.New(unexpectedSteamClosure)
}

func (w *Writer) SetMode(mode WriterMode) bool {
	if w.hasAccumulatedErr {
		return false
	}
	select {
	case <-w.responses.Outlet():
		w.hasAccumulatedErr = true
		return false
	case w.requests.Inlet() <- WriterRequest{Command: WriterSetMode, Config: WriterConfig{Mode: mode}}:
	}
	for res := range w.responses.Outlet() {
		if res.Command == WriterSetMode {
			return res.Ack
		}
	}
	w.logger.DPanic(unexpectedSteamClosure)
	return false
}

func (w *Writer) Close() (err error) {
	w.requests.Close()
	for range w.responses.Outlet() {
	}
	return w.wg.Wait()
}
